### PR TITLE
net: dhcp: Add support for restarting DHCP

### PR DESCRIPTION
**Pull request by krish2718**
If a L2 link has been established, then the DHCP is taking too long as
    it has to go through its capped exponential backoff timers to trigger
    discover (The DHCP starts immediately during init, this is itself wrong,
    it should start on a link UP notification) that delays the DHCP for
    few seconds to a minute.
    
    And if we do stop and start DHCP then also it goes through the initial
    delays (though configurable), which is also not ideal.
    
    Add support for restarting DHCP without any delay, i.e., release and
    send discover immediately.
    
    This is also useful in case L2 switches to a different subnet, in this
    case Zephyr doesn't restart DHCP automatically, this API can be used by
    L2 apps/drivers to restart DHCP to get new subnet IP.
    
    Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

### CODE REVIEW

File: subsys/net/ip/dhcpv4.c [hunk 1]
```diff
@@ -1243,6 +1243,16 @@ void net_dhcpv4_stop(struct net_if *iface)
 	k_mutex_unlock(&lock);
 }
 
+void net_dhcpv4_restart(struct net_if *iface)
+{
+
+	net_dhcpv4_stop(iface);
+
+	/* Start re-negotiation immediately */
+	dhcpv4_enter_selecting(iface);
```
[the remark this answers is not in the document]
Comment (krish2718) [reply subsys/net/ip/dhcpv4.c:1252]: You are right, I missed it.

**Comment by krish2718** [comment]
As per https://github.com/zephyrproject-rtos/zephyr/issues/49184#issuecomment-1218213533 this now looks redundant, as we can use NET_MGMT events to achieve the same functionality, so, abandoning this PR.

**Comment by krish2718** [comment]
Thinking a bit on this, we can tie up/down or carrier on/off to association/disassociation (equivalent to pluggin in cable) but that would mean DHCP client immediately starts, this is fine for Wi-Fi OPEN mode, but for secure Wi-Fi the 802.1x port is still not authorised this would mean we again enter the DHCP retries as DHCP packet will be dropped till 802.1x port is not authorized, and we cannot hold off these events till we are authorized as we use sockets to send EAPoL frames, so, even they will be dropped. So, we still might need DHCP restart?  (Re-opening for discussion).

File: subsys/net/ip/dhcpv4.c [hunk 2]
```diff
@@ -1243,6 +1243,17 @@ void net_dhcpv4_stop(struct net_if *iface)
 	k_mutex_unlock(&lock);
 }
 
+void net_dhcpv4_restart(struct net_if *iface)
+{
+	net_dhcpv4_stop(iface);
```
Comment (rlubos) [inline subsys/net/ip/dhcpv4.c:1248]: The `net_dhcpv4_stop()` function will issue a `NET_EVENT_IPV4_DHCP_STOP` event, but the restart will not issue a `NET_EVENT_IPV4_DHCP_START` afterwards. It also seems that we're missing some additional actions here (randomizing `xid`, registering net_mgmt callback).

Perhaps it'd be better to reuse the existing codebase, i.e. move `net_dhcpv4_start()` functionality into a static function, with additional `timeout` parameter, and then from `net_dhcpv4_start()` call it with a random timeout, while from `net_dhcpv4_restart()` just setting the timeout to 0?

Comment (krish2718) [reply subsys/net/ip/dhcpv4.c:1248]: Sure, makes sense.

File: subsys/net/ip/dhcpv4.c [hunk 3]
```diff
@@ -1139,10 +1139,11 @@ const char *net_dhcpv4_state_name(enum net_dhcpv4_state state)
 	return name[state];
 }
 
-void net_dhcpv4_start(struct net_if *iface)
+
```
Comment (rlubos) [inline subsys/net/ip/dhcpv4.c:1142]: No need to add extra empty line here.

File: subsys/net/ip/dhcpv4.c [hunk 4]
```diff
@@ -1166,15 +1167,17 @@ void net_dhcpv4_start(struct net_if *iface)
...
-		/* RFC2131 4.1.1 requires we wait a random period
-		 * between 1 and 10 seconds before sending the initial
-		 * discover.
-		 */
-		timeout = entropy %
-				(CONFIG_NET_DHCPV4_INITIAL_DELAY_MAX -
-				  DHCPV4_INITIAL_DELAY_MIN) +
-				DHCPV4_INITIAL_DELAY_MIN;
+		/* Use default */
+		if (delay < 0) {
```
Comment (rlubos) [inline subsys/net/ip/dhcpv4.c:1171]: I was thinking more of a providing an exact delay value in seconds in this parameter, but this apporach should also work. I would however convert it to `bool` type, IMO there's no need to use ambiguous integer values here.

Comment (krish2718) [reply subsys/net/ip/dhcpv4.c:1171]: Yeah, but as entropy is needed for both delay and XID, this seemed easier.

File: subsys/net/ip/dhcpv4.c [hunk 5]
```diff
@@ -1200,6 +1200,11 @@ void net_dhcpv4_start(struct net_if *iface)
 	k_mutex_unlock(&lock);
 }
 
+void net_dhcpv4_start(struct net_if *iface)
+{
+	return __net_dhcpv4_start(iface, 1);
```
Comment (rlubos) [inline subsys/net/ip/dhcpv4.c:1205]: nit: Please use `true`/`false`.

**Review by rlubos — DISMISSED** [review]
Thank you. LGTM

**Review by jukkar — DISMISSED** [review]

**Review by rlubos — DISMISSED** [review]

**Review by jukkar — COMMENTED** [review]
Looks good. I am not sure about the `__` prefix, at some point all the zephyr code base was changed and those prefixes were removed from the function names. So I am pondering should we use them here. What about calling the function `dhcpv4_start_internal()` or similar?

**Comment by krish2718** [comment]
> Looks good. I am not sure about the `__` prefix, at some point all the zephyr code base was changed and those prefixes were removed from the function names. So I am pondering should we use them here. What about calling the function `dhcpv4_start_internal()` or similar?

Ah okay, wasn't aware of that. Sure, we can rename.

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**